docroot/modules/humsci/hs_config_readonly/src/EventSubscriber/ConfigReadOnlyEventSubscriber.php

```diff
@@ -0,0 +1,222 @@
+
+  /**
+   * Root path of Drupal.
+   *
+   * @var string
+   */
+  protected $drupalRoot;
+
+  /**
+   * Form ids to mark as read only.
+   */
+  protected $readOnlyFormIds = [
+    'config_single_import_form',
+    'system_modules',
+    'system_modules_uninstall',
+    'user_admin_permissions',
+  ];
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct($drupal_root, ModuleHandlerInterface $module_handler, ExtensionInstallStorage $extension_config_storage, ExtensionInstallStorage $extension_optional_config_storage) {
+    $this->drupalRoot = $drupal_root;
+    $this->moduleHandler = $module_handler;
+    $this->extensionConfigStorage = $extension_config_storage;
+    $this->extensionOptionalConfigStorage = $extension_optional_config_storage;
+  }
+
+  public function onFormAlter(ReadOnlyFormEvent $event) {
+    // Check if the form is a ConfigFormBase or a ConfigEntityListBuilder.
+    $form_object = $event->getFormState()->getFormObject();
+    $mark_form_read_only = $form_object instanceof ConfigFormBase;
+
+    if (!$mark_form_read_only) {
+      $mark_form_read_only = in_array($form_object->getFormId(), $this->readOnlyFormIds);
+    }
+
+    // Check if the form is an EntityFormInterface and entity is a config
+    // entity.
+    if (!$mark_form_read_only && $form_object instanceof EntityFormInterface) {
+      $entity = $form_object->getEntity();
+      $mark_form_read_only = $entity instanceof ConfigEntityInterface;
+    }
+
+    // Don't block particular patterns.
+    if ($mark_form_read_only && $form_object instanceof EntityFormInterface) {
+      $entity = $form_object->getEntity();
+      $name = $entity->getConfigDependencyName();
+
+      // Block config from a module.
+      $mark_form_read_only = $this->configIsEditable([$name]);
+
+      // Unless its defined as a whiteist pattern.
+      if ($this->matchesWhitelistPattern($name)) {
+        $mark_form_read_only = FALSE;
+      }
+    }
+
+    // Config forms.
+    if ($mark_form_read_only && $form_object instanceof ConfigFormBase) {
+      $editable_config = $this->getEditableConfigNames($form_object);
+      $mark_form_read_only = $this->configIsEditable($editable_config);
+    }
+
+    if ($mark_form_read_only) {
+      $event->markFormReadOnly();
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents() {
+    $events = [];
+    $events[ReadOnlyFormEvent::NAME][] = ['onFormAlter', 200];
+    return $events;
+  }
+
+  /**
+   * Get the editable configuration names.
+   *
+   * @param ConfigFormBase $form
+   *   The configuration form.
+   *
+   * @return array
+   *   An array of configuration object names that are editable if called in
+   *   conjunction with the trait's config() method.
+   *
+   * @throws \ReflectionException
+   */
+  protected function getEditableConfigNames(ConfigFormBase $form) {
+    // Use reflection to work around getEditableConfigNames() as protected.
```

Comment:
Thanks for this note.

Reply:
Stole it from the original event subscriber lol

Reply:
Take note. 😆

Reply:
Note that I want to hammer this point home, I just could note help myself from posting one more note.